python/ray/data/datasource/parquet_datasource.py

```diff
@@ -52,9 +56,9 @@
 NUM_CPUS_FOR_META_FETCH_TASK = 0.5
 
 # The application-level exceptions to retry for metadata prefetching task.
-# Default to retry on `OSError` because AWS S3 would throw this transient
-# error when load is too high.
-RETRY_EXCEPTIONS_FOR_META_FETCH_TASK = [OSError]
+# Default to retry on access denied and read timeout errors because AWS S3 would throw
+# these transient errors when load is too high.
+RETRY_EXCEPTIONS_FOR_META_FETCH_TASK = ["AWS Error ACCESS_DENIED", "Timeout"]
```

Comment:
why we change from `OSError` to this now? Better to keep `OSError` to not miss any exceptions.

Reply:
Not all `OSError` are transient. How would we prevent retrying non-transient errors if we use a broad`OSError`?

Reply:
What would be non-transient errors from `OSError`? I think it's okay to retry `OSError` for now, it's okay to be over estimate here. When user report us there's job hanging / keep retrying, we can make it finer grain in future release.

Reply:
One example is rate limiting errors that don't go away. In this case, your program would hang. How would you know what's going on if we don't eventually raise an error?

Also, if there are other non-transient `OSError`s, how would you disable retries for those errors?

Reply:
> How would you know what's going on if we don't eventually raise an error?

Good question. Have we logged the error in log file? If not, we should do it. So when the program hangs, we should check the log.

> Also, if there are other non-transient OSErrors, how would you disable retries for those errors?

User can override `RETRY_EXCEPTIONS_FOR_META_FETCH_TASK` variable in their code by themselves. So I think they can work around.

Maybe we should not set `max_task_retries=-1`. We should have a upper limit, like `10` or `20`.

Reply:
> Have we logged the error in log file? If not, we should do it. So when the program hangs, we should check the log.

Don't think so. Looks like we log retries to the regular logger, not the Dataset logger.

> User can override RETRY_EXCEPTIONS_FOR_META_FETCH_TASK variable in their code by themselves. So I think they can work around.

If I don't want to retry a specific `OSError`, I'd have to remove `OSError` from `RETRY_EXCEPTIONS_FOR_META_FETCH_TASK`, right? And then I wouldn't retry any transient errors? How would I not retry a specific `OSError`?

Reply:
> Don't think so. Looks like we log retries to the regular logger, not the Dataset logger.

Let's add to the Dataset logger.

> If I don't want to retry a specific OSError, I'd have to remove OSError from RETRY_EXCEPTIONS_FOR_META_FETCH_TASK, right? And then I wouldn't retry any transient errors? How would I not retry a specific OSError?

Oh I am referring to user can do something like we did here for string matching `RETRY_EXCEPTIONS_FOR_META_FETCH_TASK = ["AWS Error ACCESS_DENIED", "Timeout"]`.

btw if you are confident that here we can cover all transient errors case from AWS, I am okay with `["AWS Error ACCESS_DENIED", "Timeout"]` for now. I want to avoid a hotfix release after GA for this issue.

Reply:
Discussed offline: we'll raise an error if the program exceeds the maximum number of retries instructing the user to configure which errors are retried and the maximum number of retries.